fish-rust/src/env/environment_impl.rs

```diff
@@ -0,0 +1,1225 @@
+    /// we avoid these copies in the common case?
+    pub fn pop(&mut self) -> Vec<WString> {
+        let popped: EnvNodeRef;
+        if let Some(next) = self.base.locals.next() {
+            popped = mem::replace(&mut self.base.locals, next);
+        } else {
+            // Exhausted the inner scopes, put back a shadowing scope.
+            if let Some(shadowed) = self.shadowed_locals.pop() {
+                popped = mem::replace(&mut self.base.locals, shadowed);
+            } else {
+                panic!("Attempt to pop last local scope")
+            }
+        }
+        let var_names = popped.borrow().env.keys().cloned().collect();
+        var_names
+    }
+
+    /// Find the first node in the chain starting at \p node which contains the given key \p key.
+    fn find_in_chain(node: &EnvNodeRef, key: &wstr) -> Option<EnvNodeRef> {
+        #[allow(clippy::manual_find)]
+        for cur in node.iter() {
+            if cur.borrow().env.contains_key(key) {
+                return Some(cur);
+            }
+        }
+        None
+    }
+
+    /// Remove a variable from the chain \p node.
+    /// Return true if the variable was found and removed.
+    fn remove_from_chain(node: &mut EnvNodeRef, key: &wstr) -> bool {
+        for cur in node.iter() {
+            let mut cur_ref = cur.borrow_mut();
+            if let Some(var) = cur_ref.env.remove(key) {
+                if var.exports() {
+                    cur_ref.changed_exported();
+                }
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Try setting \p key as an electric or readonly variable, whose value is provided by reference in \p val.
+    /// Return an error code, or NOne if not an electric or readonly variable.
+    /// \p val will not be modified upon a None return.
+    fn try_set_electric(
+        &mut self,
+        key: &wstr,
+        query: &Query,
+        val: &mut Vec<WString>,
+    ) -> Option<EnvStackSetResult> {
+        // Do nothing if not electric.
+        let ev = ElectricVar::for_name(key)?;
+
+        // If a variable is electric, it may only be set in the global scope.
+        if query.has_scope && !query.global {
+            return Some(EnvStackSetResult::ENV_SCOPE);
+        }
+
+        // If the variable is read-only, the user may not set it.
+        if query.user && ev.readonly() {
+            return Some(EnvStackSetResult::ENV_PERM);
+        }
+
+        // Be picky about exporting.
+        if query.has_export_unexport {
+            let matches = if ev.exports() {
+                query.exports
+            } else {
+                query.unexports
+            };
+            if !matches {
+                return Some(EnvStackSetResult::ENV_SCOPE);
+            }
+        }
+
+        // Handle computed mutable electric variables.
+        if key == "umask" {
+            return Some(set_umask(val));
+        } else if key == "PWD" {
```

Comment:
TIL we can compare wstr with plain string literals.
This is nice, we should probably use native string literals wherever possible.